File: regular_assignments/day_007/find_balanced_array_index/solution.go
package findBalancedArrayIndex

func Call() {
	solve([]int{2, 1, 6, 4})
	solve([]int{1, 1, 1})
	solve([]int{1, 2, 3, 4, 5, 6, 7, 8, 9})
}

func solve(arr []int) int {
	odd_prefix_sum := createOddPrefixSum(arr)
	even_prefix_sum := createEvenPrefixSum(arr)

	count := 0
	for i := 0; i < len(arr); i++ {
		sum_even := 0
		sum_odd := 0
		if i == 0 {
			sum_odd = even_prefix_sum[len(arr)-1] - even_prefix_sum[i]
			sum_even = odd_prefix_sum[len(arr)-1] - odd_prefix_sum[i]
		} else {
			sum_odd = odd_prefix_sum[i-1] + (even_prefix_sum[len(arr)-1] - even_prefix_sum[i+1-1])
			sum_even = even_prefix_sum[i-1] + (odd_prefix_sum[len(arr)-1] - odd_prefix_sum[i+1-1])
		}
		if sum_even == sum_odd {
			count++
		}
	}
	return count
}

func createEvenPrefixSum(arr []int) []int {
	sum := []int{arr[0]}
	for i := 1; i < len(arr); i++ {
		if i%2 == 0 {
			sum = append(sum, sum[i-1]+arr[i])
		} else {
			sum = append(sum, sum[i-1])
		}
	}
	return sum
}

func createOddPrefixSum(arr []int) []int {
	sum := []int{0}
	for i := 1; i < len(arr); i++ {
		if i%2 != 0 {
			sum = append(sum, sum[i-1]+arr[i])
		} else {
			sum = append(sum, sum[i-1])
		}
	}
	return sum
}

File: regular_assignments/day_007/find_balanced_array_index/solution.rb
class Solution
  def call
    solve([2, 1, 6, 4]) # 1
    solve([1, 1, 1]) # 3
    solve([1, 2, 3, 4, 5, 6, 7, 8, 9]) # 0
  end

  def solve(arr)
    prefix_even_sum_arr = create_even_prefix_sum_arr(arr)
    prefix_odd_sum_arr = create_odd_prefix_sum_arr(arr)
    i = 0
    count = 0
    while i < arr.length
      if i.zero?
        sum_odd = prefix_even_sum_arr[arr.length - 1] - prefix_even_sum_arr[i]
        sum_even = prefix_odd_sum_arr[arr.length - 1] - prefix_odd_sum_arr[i]
      else
        sum_odd = prefix_odd_sum_arr[i - 1] + (prefix_even_sum_arr[arr.length-1] - prefix_even_sum_arr[i-1+1])
        sum_even = prefix_even_sum_arr[i - 1] + (prefix_odd_sum_arr[arr.length-1] - prefix_odd_sum_arr[i-1+1])
      end
      count += 1 if sum_even == sum_odd
      i += 1
    end

    count
  end

  def create_odd_prefix_sum_arr(arr)
    i = 1
    prefix_sum = [0]
    while i < arr.length
      prefix_sum[i] = !(i % 2).zero? ? prefix_sum[i - 1] + arr[i] : prefix_sum[i - 1]
      i += 1
    end
    prefix_sum
  end

  def create_even_prefix_sum_arr(arr)
    i = 1
    prefix_sum = [arr[0]]
    while i < arr.length
      prefix_sum[i] = (i % 2).zero? ? prefix_sum[i - 1] + arr[i] : prefix_sum[i - 1]
      i += 1
    end
    prefix_sum
  end
end

# puts '-' * 100
# puts Solution.new.solve([2, 1, 6, 4])
# puts '-' * 100
# puts Solution.new.solve([1, 1, 1])
# puts '-' * 100
# puts Solution.new.solve([1, 2, 3, 4, 5, 6, 7, 8, 9])
